Docker secrets files used as string literals

**Describe the bug**
When setting 

````yml
      MB_DB_USER: /run/secrets/db_user
      MB_DB_PASS: /run/secrets/db_password
      MB_DB_DBNAME: /run/secrets/db_name
````

as per the  documentation in a docker compose file, with the appropriate secret configuration, the literal  "/run/secrets/db_name" , "/run/secrets/db_user" and "/run/secrets/db_password" are used as the database name,  user and password.

````cli
metabase_1        | Caused by: java.sql.SQLInvalidAuthorizationSpecException: (conn=405) Access denied for user '/run/secrets/db_user'@'172.19.0.16' (using password: YES)
metabase_1        | Current charset is UTF-8. If password has been set using other charset, consider using option 'passwordCharacterEncoding'
````

Suggested fix would be to follow the postgres and mysql container secrets environment variable conventions.
They append _FILE to the environment variable if it is populated using a secret.
@see https://hub.docker.com/_/Mysql
@see https://hub.docker.com/_/postgres

Example 
````yml
      MB_DB_USER: /run/secrets/db_user
      MB_DB_PASS: /run/secrets/db_password
      MB_DB_DBNAME: /run/secrets/db_name
````

Would become 
````yml
      MB_DB_USER_FILE: /run/secrets/db_user
      MB_DB_PASS_FILE: /run/secrets/db_password
      MB_DB_DBNAME_FILE: /run/secrets/db_name
````

These changes could be implemented with a small mod to the /bin/docker/run_metabase.sh file.

````bash
# check for docker secrets file
if [[ -n "${MB_DB_USER_FILE}" ]]; then
  export MB_DB_USER=$(cat ${MB_DB_USER_FILE})
  echo "metabase db username set from secrets file ${MB_DB_USER_FILE}"
fi

if [[  -n "${MB_DB_PASS_FILE}" ]]; then
    export MB_DB_PASS=$(cat ${MB_DB_PASS_FILE})
  echo "metabase db password set from secrets file ${MB_DB_PASS_FILE}"
fi

if [[  -n "${MB_DB_DBNAME_FILE}" ]]; then
    export MB_DB_DBNAME=$(cat ${MB_DB_DBNAME_FILE})
  echo "metabase db name set from secrets file ${MB_DB_DBNAME_FILE}"
fi
````

Well, @dannydnz  opened the issue to tell that the values in MB_DB_USER and MB_DB_PASS were being treated as strings, as you also said.

The point is that this page https://www.metabase.com/docs/latest/operations-guide/running-metabase-on-docker.html shows that we can use secrets at MB_DB_USER and MB_DB_PASS, buts this simply do not work.

My docker-compose.yml is:

```yml
version: '3.8'

secrets:
  mariadb_root_password:
    external: true
  metabase_database_user:
    external: true
  metabase_database_password:
    external: true

services:	
  database:
    image: mariadb:10.6.4
    secrets:
      - mariadb_root_password
      - metabase_database_user
      - metabase_database_password
    environment:
      MARIADB_ROOT_PASSWORD_FILE: /run/secrets/mariadb_root_password
      MARIADB_USER_FILE: /run/secrets/metabase_database_user
      MARIADB_PASSWORD_FILE: /run/secrets/metabase_database_password
      MARIADB_DATABASE: ${METABASE_DATABASE_NAME}

  metabase:
    image: metabase/metabase:v0.40.3.1
    depends_on: 
      - database
    networks:
      - metabase
    ports:
      - 3000:3000
    secrets:
      - metabase_database_user
      - metabase_database_password
    environment:
      MB_DB_TYPE: mysql
      MB_DB_HOST: database
      MB_DB_PORT: 3306
      MB_DB_USER: /run/secrets/metabase_database_user
      MB_DB_PASS: /run/secrets/metabase_database_password
      MB_DB_DBNAME: ${METABASE_DATABASE_NAME}
      JAVA_TIMEZONE: Europe/London
```

The error is always `Caused by: java.sql.SQLInvalidAuthorizationSpecException: Could not connect to address=(host=database)(port=3306)(type=master) : (conn=19) Access denied for user '/run/secrets/metabase_database_user'@'10.0.14.6' (using password: YES)`

The conclusion is:

1. As you said, the two mentioned variables **need to be string**, so the doc cannot show tha a secret can be used.
2. This issue is a feature request.
3. The doc has an error caused by #14839

Tested with:

```yml
secrets:
  mariadb_root_password:
    file: ./mariadb_root_password.secret
  metabase_database_user:
    file: ./metabase_database_user.secret
  metabase_database_password:
    file: ./metabase_database_password.secret
```
and
```yml
secrets:
  mariadb_root_password:
    external: true
  metabase_database_user:
    external: true
  metabase_database_password:
    external: true
```

In both cases, mariadb environment variables (see previous comment for details) that uses secrets worked, but not the ones from metabase, with the error `Caused by: java.sql.SQLInvalidAuthorizationSpecException: (conn=3) Access denied for user '/run/secrets/metabase_database_user'@'10.0.19.4' (using password: YES)`

I am confident to say that the metabase image do not supports docker secrets and the examples using it should be removed from the documentation.

--- a/bin/docker/run_metabase.sh
+++ b/bin/docker/run_metabase.sh
@@ -16,6 +16,22 @@ if [ ! -z "$JAVA_TIMEZONE" ]; then
     JAVA_OPTS="${JAVA_OPTS} -Duser.timezone=${JAVA_TIMEZONE}"
 fi
 
+# check for docker secrets file
+if [[ -n "${MB_DB_USER_FILE}" ]]; then
+  export MB_DB_USER=$(cat ${MB_DB_USER_FILE})
+  echo "metabase db username set from secrets file ${MB_DB_USER_FILE}"
+fi
+
+if [[  -n "${MB_DB_PASS_FILE}" ]]; then
+    export MB_DB_PASS=$(cat ${MB_DB_PASS_FILE})
+  echo "metabase db password set from secrets file ${MB_DB_PASS_FILE}"
+fi
+
+if [[  -n "${MB_DB_DBNAME_FILE}" ]]; then
+    export MB_DB_DBNAME=$(cat ${MB_DB_DBNAME_FILE})
+  echo "metabase db name set from secrets file ${MB_DB_DBNAME_FILE}"
+fi
+
 # detect if the container is started as root or not
 # if non-root, it's likely we run in a k8s environment with well maintained permissions
 # if root, we need to check some permissions in order to exec metabase with a non-root user

--- a/bin/docker/run_metabase.sh
+++ b/bin/docker/run_metabase.sh
@@ -16,21 +16,33 @@ if [ ! -z "$JAVA_TIMEZONE" ]; then
     JAVA_OPTS="${JAVA_OPTS} -Duser.timezone=${JAVA_TIMEZONE}"
 fi
 
-# check for docker secrets file
-if [[ -n "${MB_DB_USER_FILE}" ]]; then
-  export MB_DB_USER=$(cat ${MB_DB_USER_FILE})
-  echo "metabase db username set from secrets file ${MB_DB_USER_FILE}"
-fi
+file_env() {
+	local var="$1"
+	local fileVar="${var}_FILE"
+	local def="${2:-}"
+	if [ "${!var:-}" ] && [ "${!fileVar:-}" ]; then
+		echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
+		exit 1
+	fi
+	local val="$def"
+	if [ "${!var:-}" ]; then
+		val="${!var}"
+	elif [ "${!fileVar:-}" ]; then
+		val="$(< "${!fileVar}")"
+	fi
+	export "$var"="$val"
+	unset "$fileVar"
+}
 
-if [[  -n "${MB_DB_PASS_FILE}" ]]; then
-    export MB_DB_PASS=$(cat ${MB_DB_PASS_FILE})
-  echo "metabase db password set from secrets file ${MB_DB_PASS_FILE}"
-fi
-
-if [[  -n "${MB_DB_DBNAME_FILE}" ]]; then
-    export MB_DB_DBNAME=$(cat ${MB_DB_DBNAME_FILE})
-  echo "metabase db name set from secrets file ${MB_DB_DBNAME_FILE}"
-fi
+docker_setup_env() {
+    file_env 'MB_DB_USER'
+    file_env 'MB_DB_PASS'
+    file_env 'MB_DB_CONNECTION_URI'
+    file_env 'MB_EMAIL_SMTP_PASSWORD'
+    file_env 'MB_EMAIL_SMTP_USERNAME'
+    file_env 'MB_LDAP_PASSWORD'
+    file_env 'MB_LDAP_BIND_DN'
+}
 
 # detect if the container is started as root or not
 # if non-root, it's likely we run in a k8s environment with well maintained permissions
@@ -40,6 +52,7 @@ if [ $(id -u) -ne 0 ]; then
     # Launch the application
     # exec is here twice on purpose to  ensure that metabase runs as PID 1 (the init process)
     # and thus receives signals sent to the container. This allows it to shutdown cleanly on exit
+    docker_setup_env
     exec /bin/sh -c "exec java $JAVA_OPTS -jar /app/metabase.jar $@"
 else
     # Avoid running metabase (or any server) as root where possible
@@ -120,6 +133,7 @@ else
 
     # next we tell metabase use the files we just moved into the directory
     # or create the files in that directory if they don't exist.
+    docker_setup_env
     export MB_DB_FILE=$new_db_dir/$(basename $db_file)
 
     # TODO: print big scary warning if they are configuring an ephemeral instance

--- a/bin/docker/run_metabase.sh
+++ b/bin/docker/run_metabase.sh
@@ -16,6 +16,12 @@ if [ ! -z "$JAVA_TIMEZONE" ]; then
     JAVA_OPTS="${JAVA_OPTS} -Duser.timezone=${JAVA_TIMEZONE}"
 fi
 
+# usage: file_env VAR [DEFAULT]
+#    ie: file_env 'XYZ_DB_PASSWORD' 'example'
+# (will allow for "$XYZ_DB_PASSWORD_FILE" to fill in the value of
+#  "$XYZ_DB_PASSWORD" from a file, especially for Docker's secrets feature)
+# taken from https://github.com/docker-library/postgres/blob/master/docker-entrypoint.sh
+# This is the specific function that takes the env var which has a "_FILE" at the end and transforms that into a normal env var.
 file_env() {
 	local var="$1"
 	local fileVar="${var}_FILE"
@@ -34,6 +40,7 @@ file_env() {
 	unset "$fileVar"
 }
 
+# Here we define which env vars are the ones that will be supported with a "_FILE" ending. We started with the ones that would contain sensitive data
 docker_setup_env() {
     file_env 'MB_DB_USER'
     file_env 'MB_DB_PASS'
@@ -48,6 +55,7 @@ docker_setup_env() {
 # if non-root, it's likely we run in a k8s environment with well maintained permissions
 # if root, we need to check some permissions in order to exec metabase with a non-root user
 # In that case, the container is run as root, metabase is run as a non-root user
+# Also, we call the docker_setup_env function before Metabase starts so it takes the Docker secrets in case there are any
 if [ $(id -u) -ne 0 ]; then
     # Launch the application
     # exec is here twice on purpose to  ensure that metabase runs as PID 1 (the init process)

--- a/docs/operations-guide/running-metabase-on-docker.md
+++ b/docs/operations-guide/running-metabase-on-docker.md
@@ -161,7 +161,7 @@ Note that Metabase will use this directory to extract plugins bundled with the d
 
 In order to keep your connection parameters hidden from plain sight, you can use Docker Secrets to put all parameters in files so Docker can read and load them in memory before the container is started.
 
-This is an example of a `docker-compose.yml` file to start a Metabase container with secrets to connect to a PostgreSQL database. Create 2 files (db_user.txt and db_password.txt) in the same directory as this `yml` and fill them with any username and a secure password:
+This is an example of a `docker-compose.yml` file to start a Metabase container with secrets to connect to a PostgreSQL database. Create 2 files (db_user.txt and db_password.txt) in the same directory as this `yml` and fill them with any username and a secure password (notice the "_FILE" on the environment variables that have a secret):
 
 ```
 version: '3.9'
@@ -178,8 +178,8 @@ services:
       MB_DB_TYPE: postgres
       MB_DB_DBNAME: metabase
       MB_DB_PORT: 5432
-      MB_DB_USER: /run/secrets/db_user
-      MB_DB_PASS: /run/secrets/db_password
+      MB_DB_USER_FILE: /run/secrets/db_user
+      MB_DB_PASS_FILE: /run/secrets/db_password
       MB_DB_HOST: postgres-secrets
     networks: 
       - metanet1-secrets
@@ -193,9 +193,9 @@ services:
     container_name: postgres-secrets
     hostname: postgres-secrets
     environment:
-      POSTGRES_USER: /run/secrets/db_user
+      POSTGRES_USER_FILE: /run/secrets/db_user
       POSTGRES_DB: metabase
-      POSTGRES_PASSWORD: /run/secrets/db_password
+      POSTGRES_PASSWORD_FILE: /run/secrets/db_password
     networks: 
       - metanet1-secrets
     secrets:
@@ -210,3 +210,15 @@ secrets:
    db_user:
      file: db_user.txt
 ```
+
+We currently support the following [environment variables](environment-variables.html) to be used as secrets:
+
+* MB_DB_USER
+* MB_DB_PASS
+* MB_DB_CONNECTION_URI
+* MB_EMAIL_SMTP_PASSWORD
+* MB_EMAIL_SMTP_USERNAME
+* MB_LDAP_PASSWORD
+* MB_LDAP_BIND_DN
+
+In order for the Metabase container to read the files and use the contents as a secret, the environment variable name needs to be appended with a "_FILE" as explained above.

